### PR TITLE
feat: Qdrant support

### PR DESCRIPTION
**Pull request by Anush008**
Adds factory implementation to pick a vector store.
- Qdrant
- Milvus(Default)

~~I see there's been changes to the implementation upstream. I'll update this PR.~~ Done in [`5151530` (#19)](https://github.com/whyhow-ai/knowledge-table/pull/19/commits/51515300bf74b46817e60ee5efb791103c663030)

~~I am QA testing right now.~~ 🟢

To run Qdrant

```bash
docker run -p 6333:6333 qdrant/qdrant
````

You'll find the dashboard at http://localhost:6333/dashboard

### CODE REVIEW

File: backend/.env.sample [hunk 1]
```diff
@@ -1,25 +1,50 @@
 # -------------------------
 # LLM CONFIG
 # -------------------------
-DIMENSIONS=768
-EMBEDDING_PROVIDER=openai
+
 LLM_PROVIDER=openai
-OPENAI_API_KEY={your-openai-api-key}
```
Comment (yaserabdelaziz) [inline backend/.env.sample:7]: Why was this line removed when `OPENAI_API_KEY` is still being used?

File: backend/.env.sample [hunk 2]
```diff
@@ -1,25 +1,50 @@
...
 INDEX_NAME={your-index-name}
+EMBEDDING_PROVIDER=openai
+DIMENSIONS=768
+
+# -------------------------
+# MILVUS CONFIG
+# Applicable if VECTOR_DB=milvus-lite
+# -------------------------
 MILVUS_DB_USERNAME=root
-MILVUS_DB_PASSWORD={your-milvus-password}
```
Comment (yaserabdelaziz) [inline backend/.env.sample:15]: Why was this line removed when `MILVUS_DB_PASSWORD` is still being used?

File: backend/src/knowledge_table_api/config.py [hunk 3]
```diff
@@ -29,7 +29,7 @@ class Settings(BaseSettings):
     unstructured_api_key: str | None = None
 
     model_config = SettingsConfigDict(
-        env_file=".env", env_file_encoding="utf-8"
+        env_file=".env", env_file_encoding="utf-8", extra="allow"
```
Comment (yaserabdelaziz) [inline backend/src/knowledge_table_api/config.py:32]: It would be safer to avoid using `extra="allow"` to ensure that all configuration fields are explicitly defined. This can help prevent potential misconfigurations or unexpected behavior due to unrecognized fields in the environment file. Please, consider removing `extra="allow"` from the `SettingsConfigDict` to enforce strict validation of configuration fields.

File: backend/src/knowledge_table_api/services/vector_index/qdrant.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+                seen_chunks.add(chunk["chunk_number"])
+                formatted_output.append(
+                    {"content": chunk["text"], "page": chunk["page_number"]}
+                )
+
+        logger.info(f"Retrieved {len(formatted_output)} unique chunks.")
+        return VectorResponse(
+            message="Query processed successfully.",
+            chunks=[Chunk(**chunk) for chunk in formatted_output],
+        )
```
Comment (yaserabdelaziz) [inline backend/src/knowledge_table_api/services/vector_index/qdrant.py:113]: It looks like the `VectorResponse` class is used as the return type, but the return type annotation is `dict[str, Any]`. To maintain consistency and avoid potential type errors, consider changing the return statement to match the `dict[str, Any]` annotation. This will ensure the code accurately reflects the intended return type and aligns with parent class type annotations.

File: backend/src/knowledge_table_api/services/vector_index/base.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    @abstractmethod
+    async def upsert_vectors(
+        self, document_id: str, chunks: List[Document], llm_service: LLMService
+    ) -> Dict[str, str]:
+        pass
+
+    @abstractmethod
+    async def vector_search(
+        self, queries: List[str], document_id: str, llm_service: LLMService
+    ) -> dict[str, Any]:
```
Comment (yaserabdelaziz) [inline backend/src/knowledge_table_api/services/vector_index/base.py:33]: To ensure consistency, it makes sense to consider returning `VectorResponse` here. This change would provide a structured and validated response, enhancing code readability and type safety.

File: backend/src/knowledge_table_api/services/vector_index/base.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        pass
+
+    @abstractmethod
+    async def decomposed_search(
+        self,
+        query: str,
+        document_id: str,
+        rules: List[Rule],
+        llm_service: LLMService,
+    ) -> Dict[str, Any]:
```
Comment (yaserabdelaziz) [inline backend/src/knowledge_table_api/services/vector_index/base.py:53]: To ensure consistency, it makes sense to consider returning `VectorResponse` here. This change would provide a structured and validated response, enhancing code readability and type safety.

**Review by yaserabdelaziz — CHANGES_REQUESTED** [review]
LGTM. I left a few minor comments, particularly about adhering to type annotations.

**Comment by Anush008** [comment]
Wow. The repo just got a refactor again.
I'm updating my changes.

**Comment by tomsmoker** [comment]
@Anush008 no need. Your work is good, I'll integrate. Thanks.

**Comment by Anush008** [comment]
> @Anush008 no need. Your work is good, I'll integrate. Thanks.

I was done already. There they are.

File: backend/src/app/services/vector_db/base.py [hunk 7]
```diff
@@ -58,9 +77,85 @@ async def ensure_collection_exists(self) -> None:
         """Ensure the collection exists in the vector database."""
         pass
 
-    @abstractmethod
+    async def get_embeddings(
```
Comment (Anush008) [inline backend/src/app/services/vector_db/base.py:80]: I have moved these methods to the base class because they're shared by both Qdrant and Milvus.

Comment (tomsmoker) [reply backend/src/app/services/vector_db/base.py:80]: Awesome! Thanks

**Comment by tomsmoker** [comment]
@Anush008 up to date, Milvus PR merged.  Want me to update this one? Won't be much logical change, mostly just formatting.

**Comment by tomsmoker** [comment]
@Anush008 done, all passing, just needs a changelog update.

**Review by tomsmoker — APPROVED** [review]
Have added my changes.  Looks great.  Thanks @Anush008
